File: python/hyperspace/hyperspace.py
import sys

from pyspark.sql import SparkSession, DataFrame
from py4j.java_gateway import java_import

from .indexconfig import *


class Hyperspace:
    def __init__(self, spark):
        """
        Initializes Hyperspace object.
        :param spark: sparkSession object
        :return: Hyperspace object

        >>> hyperspace = Hyperspace(spark)
        """
        self.spark = spark
        self.jvm = spark._jvm
        self.hyperspace = self.jvm.com.microsoft.hyperspace.Hyperspace(spark._jsparkSession)

    def _getJavaIndexConfig(self, index_config):
        """
        Constructs IndexConfig Java object from python wrapper IndexConfig object.
        :param index_config: IndexConfig java object
        :return: IndexConfig python object

        >>> _getJavaIndexConfig(idx_config)
        """
        indexed_columns = self._getScalaSeqFromList(index_config.indexedColumns)
        included_columns = self._getScalaSeqFromList(index_config.includedColumns)
        _jindexConfig = self.jvm.com.microsoft.hyperspace.index.IndexConfig(
            self.jvm.java.lang.String(index_config.indexName), indexed_columns, included_columns)
        return _jindexConfig

    def _getScalaSeqFromList(self, list):
        """
        Constructs scala sequence from Java's List object.
        :param list: List object in Java
        :return: Seq object in scala

        >>> _getScalaSeqFromList(list)
        """
        java_import(self.jvm, "scala.collection.JavaConversions._")
        java_import(self.jvm, "scala.collection.Seq")
        java_import(self.jvm, 'java.util.*')
        result_array_list = self.jvm.ArrayList(len(list))
        for element in list:
            result_array_list.add(self.jvm.String(element))
        return self.jvm.scala.collection.JavaConverters.asScalaIteratorConverter(
            result_array_list.iterator()).asScala().toSeq()

    def indexes(self):
        """
        Gets available indexes.
        :return: dataFrame object containing list of indexes.

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.indexes()
        """
        return DataFrame(self.hyperspace.indexes(), self.spark._wrapped)

    def createIndex(self, dataFrame, indexConfig):
        """
        Creates index on the given dataframe using the given indexConfig.
        :param dataFrame: dataFrame
        :param indexConfig: indexConfig

        >>> hyperspace = Hyperspace(spark)
        >>> idxConfig = IndexConfig("indexName", ["c1"], ["c2","c3"])
        >>> df = spark.read.parquet("./sample.parquet").toDF("c1", "c2", "c3")
        >>> hyperspace.createIndex(df, indexConfig)
        """
        self.hyperspace.createIndex(dataFrame._jdf, self._getJavaIndexConfig(indexConfig))

    def deleteIndex(self, indexName):
        """
        Soft deletes given index.
        :param indexName: index name

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.deleteIndex("indexname")
        """
        self.hyperspace.deleteIndex(indexName)

    def restoreIndex(self, indexName):
        """
        Restores index with given index name.
        :param indexName: index name

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.restoreIndex("indexname")
        """
        self.hyperspace.restoreIndex(indexName)

    def vacuumIndex(self, indexName):
        """
        Vacuums index with given index name.
        :param indexName: index name

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.vacuumIndex("indexname")
        """
        self.hyperspace.vacuumIndex(indexName)

    def refreshIndex(self, indexName, mode='full'):
        """
        Update indexes for the latest version of the data.
        :param indexName: index name
        :param mode: refresh mode 'incremental' or 'full'

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.refreshIndex("indexname", "incremental/full")
        """
        self.hyperspace.refreshIndex(indexName, mode)

    def optimizeIndex(self, indexName, mode='quick'):
        """
        Optimize index by changing the underlying index data layout (e.g., compaction).
        :param indexName: index name
        :param mode: optimize mode 'quick' or 'full'

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.optimizeIndex("indexname", "quick/full")
        """
        self.hyperspace.optimizeIndex(indexName, mode)

    def cancel(self, indexName):
        """
        Cancel api to bring back index from an inconsistent state to the last known stable state.
        :param indexName: index name

        >>> hyperspace = Hyperspace(spark)
        >>> hyperspace.cancel("indexname")
        """
        self.hyperspace.cancel(indexName)

    def explain(self, df, verbose=False, redirectFunc=lambda x: sys.stdout.write(x)):
        """
        Explains how indexes will be applied to the given dataframe.
        :param df: dataFrame
        :param redirectFunc: optional function to redirect output of explain

        >>> hyperspace = Hyperspace(spark)
        >>> df = spark.read.parquet("./sample.parquet").toDF("c1", "c2", "c3")
        >>> hyperspace.explain(df)
        """
        analyzer = self.jvm.com.microsoft.hyperspace.index.plananalysis.PlanAnalyzer
        result_string = analyzer.explainString(df._jdf, self.spark._jsparkSession,
                                               self.hyperspace.indexes(), verbose)
        redirectFunc(result_string)

    @staticmethod
    def enable(spark):
        """
        Enables Hyperspace index usage on given spark session.
        :param spark: sparkSession

        >>> Hyperspace.enable(spark)
        """
        spark._jvm.com.microsoft.hyperspace.util.PythonUtils.enableHyperspace(spark._jsparkSession)
        return spark

    @staticmethod
    def disable(spark):
        """
        Disables Hyperspace index usage on given spark session.
        :param spark: sparkSession

        >>> Hyperspace.disable(spark)
        """
        spark._jvm.com.microsoft.hyperspace.util.PythonUtils.disableHyperspace(spark._jsparkSession)
        return spark

    @staticmethod
    def isEnabled(spark):
        """
        Checks if Hyperspace is enabled or not.
        :param spark: sparkSession

        >>> Hyperspace.isEnabled(spark)
        """
        return spark._jvm.com.microsoft.hyperspace.util.PythonUtils. \
            isHyperspaceEnabled(spark._jsparkSession)

File: src/main/scala/com/microsoft/hyperspace/Hyperspace.scala
/*
 * Copyright (2020) The Hyperspace Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.microsoft.hyperspace

import org.apache.spark.sql.{DataFrame, SparkSession}

import com.microsoft.hyperspace.index._
import com.microsoft.hyperspace.index.IndexConstants.{OPTIMIZE_MODE_QUICK, REFRESH_MODE_FULL}
import com.microsoft.hyperspace.index.plananalysis.PlanAnalyzer

class Hyperspace(spark: SparkSession) {
  private val indexManager: IndexManager = Hyperspace.getContext(spark).indexCollectionManager

  /**
   * Collect all the index metadata.
   *
   * @return all index metadata as a [[DataFrame]].
   */
  def indexes: DataFrame = indexManager.indexes

  /**
   * Create index.
   *
   * @param df the DataFrame object to build index on.
   * @param indexConfig the configuration of index to be created.
   */
  def createIndex(df: DataFrame, indexConfig: IndexConfig): Unit = {
    indexManager.create(df, indexConfig)
  }

  /**
   * Soft deletes the index with given index name.
   *
   * @param indexName the name of index to delete.
   */
  def deleteIndex(indexName: String): Unit = {
    indexManager.delete(indexName)
  }

  /**
   * Restores index with given index name.
   *
   * @param indexName Name of the index to restore.
   */
  def restoreIndex(indexName: String): Unit = {
    indexManager.restore(indexName)
  }

  /**
   * Does hard delete of indexes marked as `DELETED`.
   *
   * @param indexName Name of the index to restore.
   */
  def vacuumIndex(indexName: String): Unit = {
    indexManager.vacuum(indexName)
  }

  /**
   * Update indexes for the latest version of the data.
   *
   * @param indexName Name of the index to refresh.
   */
  def refreshIndex(indexName: String): Unit = {
    refreshIndex(indexName, REFRESH_MODE_FULL)
  }

  /**
   * Update indexes for the latest version of the data. This API provides a few supported
   * refresh modes as listed below.
   *
   * @param indexName Name of the index to refresh.
   * @param mode Refresh mode. Currently supported modes are `incremental` and `full`.
   */
  def refreshIndex(indexName: String, mode: String): Unit = {
    indexManager.refresh(indexName, mode)
  }

  /**
   * Optimize index by changing the underlying index data layout (e.g., compaction).
   *
   * Note: This API does NOT refresh (i.e. update) the index if the underlying data changes. It
   * only rearranges the index data into a better layout, by compacting small index files. The
   * index files larger than a threshold remain untouched to avoid rewriting large contents.
   *
   * `Quick` optimize mode is used by default.
   *
   * @param indexName Name of the index to optimize.
   */
  def optimizeIndex(indexName: String): Unit = {
    indexManager.optimize(indexName, OPTIMIZE_MODE_QUICK)
  }

  /**
   * Optimize index by changing the underlying index data layout (e.g., compaction).
   *
   * Note: This API does NOT refresh (i.e. update) the index if the underlying data changes. It
   * only rearranges the index data into a better layout, by compacting small index files. The
   * index files larger than a threshold remain untouched to avoid rewriting large contents.
   *
   * Available modes:
   * `Quick` mode: This mode allows for fast optimization. Files smaller than a
   * predefined threshold "spark.hyperspace.index.optimize.fileSizeThreshold" will be picked
   * for compaction.
   *
   * `Full` mode: This allows for slow but complete optimization. ALL index files are
   * picked for compaction.
   *
   * @param indexName Name of the index to optimize.
   * @param mode Index optimization mode. "quick" refers to optimization of only small index
   *             files, based on a threshold. "full" refers to recreation of index.
   */
  def optimizeIndex(indexName: String, mode: String): Unit = {
    indexManager.optimize(indexName, mode)
  }

  /**
   * Cancel API to bring back index from an inconsistent state to the last known stable state.
   * E.g. if index fails during creation, in "CREATING" state.
   * The index will not allow any index modifying operations unless a cancel is called.
   *
   * Note: Cancel from "VACUUMING" state will move it forward to "DOESNOTEXIST" state.
   * Note: If no previous stable state exists, cancel will move it to "DOESNOTEXIST" state.
   *
   * @param indexName Name of the index to cancel.
   */
  def cancel(indexName: String): Unit = {
    indexManager.cancel(indexName)
  }

  /**
   * Explains how indexes will be applied to the given dataframe.
   *
   * @param df dataFrame.
   * @param redirectFunc optional function to redirect output of explain.
   * @param verbose Flag to enable verbose mode.
   */
  def explain(df: DataFrame, verbose: Boolean = false)(
      implicit redirectFunc: String => Unit = print): Unit = {
    redirectFunc(PlanAnalyzer.explainString(df, spark, indexManager.indexes, verbose))
  }
}

object Hyperspace {
  private lazy val context = new ThreadLocal[HyperspaceContext]

  private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
    val ctx = context.get()
    if (ctx == null || !ctx.spark.equals(spark)) {
      // Ensure the Spark session the caller provides is the same as
      // the one HyperspaceContext is using because Hyperspace depends on the
      // session's properties such as configs, etc.
      context.set(new HyperspaceContext(spark))
    }

    context.get()
  }

  def apply(): Hyperspace = {
    val sparkSession = SparkSession.getActiveSession.getOrElse {
      throw HyperspaceException("Could not find active SparkSession.")
    }

    new Hyperspace(sparkSession)
  }
}

private[hyperspace] class HyperspaceContext(val spark: SparkSession) {
  val indexCollectionManager = CachingIndexCollectionManager(spark)
}
